### PR TITLE
Make compaction not generate empty segment

### LINKED ISSUE #15630: [Bug]: Empty segment generated by compaction is still saved
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: v2.0.0
- Deployment mode(standalone or cluster):
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

A compaction may generate an empty segment because of all entities in segments are deleted. Now datacoord will just add a new segment and may compact this segment later. It's not resonable. 

### Expected Behavior

We can just ignore the empty segment generated by compaction.

### Steps To Reproduce

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by yah01**
Signed-off-by: yah01 <yah2er0ne@outlook.com>
Fix #15630

### CODE REVIEW

File: internal/datanode/compactor.go [hunk 1]
```diff
@@ -466,17 +466,23 @@ func (t *compactionTask) compact() error {
 	}
 
 	uploadStart := time.Now()
-	cpaths, err := t.upload(ctxTimeout, targetSegID, partID, iDatas, deltaBuf.delData, meta)
-	if err != nil {
-		log.Error("compact wrong", zap.Int64("planID", t.plan.GetPlanID()), zap.Error(err))
-		return err
+	var segPaths *cpaths
+	if numRows <= 0 {
```
Comment (congqixia) [inline internal/datanode/compactor.go:470]: According to the context, this line shall be 
```
if numRows > 0 {
```

File: internal/datanode/compactor.go [hunk 2]
```diff
@@ -466,17 +466,23 @@ func (t *compactionTask) compact() error {
...
+		segPaths, err = t.upload(ctxTimeout, targetSegID, partID, iDatas, deltaBuf.delData, meta)
+		if err != nil {
+			log.Error("compact wrong", zap.Int64("planID", t.plan.GetPlanID()), zap.Error(err))
+			return err
+		}
```
Comment (XuanYang-cn) [inline internal/datanode/compactor.go:475]: I don't see the point of this, `upload` already takes care of empty segment cases https://github.com/milvus-io/milvus/blob/b0923f1299fea669380456e8175f971576d244af/internal/datanode/binlog_io.go#L125

File: internal/datanode/compactor_test.go [hunk 3]
```diff
@@ -473,6 +478,17 @@ func TestCompactorInterfaceMethods(t *testing.T) {
...
+		// Compact to delete the entire segment
+		err = mockKv.RemoveWithPrefix("/")
+		require.NoError(t, err)
+		cpaths, err = mockbIO.upload(context.TODO(), segID, partID, []*InsertData{iData}, deleteAllData, meta)
+		require.NoError(t, err)
+		plan.PlanID++
+
+		err = task.compact()
+		assert.NoError(t, err)
+
```
Comment (XuanYang-cn) [inline internal/datanode/compactor_test.go:490]: Need to assert for zero num of rows.

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: internal/datanode/compactor.go (hunk 2 above)
Comment (yah01) [reply internal/datanode/compactor.go]: it just warns that you are trying to upload empty data, but still creates the blob

**Comment by congqixia** [comment]
@yah01 
Another concern is that after empty segment is ignored in `DataCoord`, what's the behavior of hand-off (with empty target segment) in query cluster?  My first guess is that the hand-off have some error if `QueryCoord` attempts to fetch the segment information with the `Dropped` segment.

File: internal/datacoord/meta.go [hunk 4]
```diff
@@ -793,6 +799,17 @@ func (m *meta) CompleteInnerCompaction(segmentBinlogs *datapb.CompactionSegmentB
 	defer m.Unlock()
 
 	if segment := m.segments.GetSegment(segmentBinlogs.SegmentID); segment != nil {
+		// The compaction deletes the entire segment
+		if result.NumOfRows <= 0 {
```
Comment (sunby) [inline internal/datacoord/meta.go:803]: Add ut to test the function when result.NumOfRows == 0 please

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**